controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -407,6 +466,55 @@ func (r *KubeadmControlPlaneReconciler) reconcileKubeconfig(ctx context.Context,
 	return nil
 }
 
+func getMachines(ctx context.Context, crClient client.Client, clusterName types.NamespacedName) (*clusterv1.MachineList, error) {
+	selector := generateKubeadmControlPlaneLabels(clusterName.Name)
+	allMachines := &clusterv1.MachineList{}
+	if err := crClient.List(ctx, allMachines, client.InNamespace(clusterName.Namespace), client.MatchingLabels(selector)); err != nil {
+		return nil, errors.Wrap(err, "failed to list machines")
+	}
+	return allMachines, nil
+}
+
+func getOwnedMachines(ctx context.Context, crClient client.Client, kcp *controlplanev1.KubeadmControlPlane, clusterName types.NamespacedName) ([]*clusterv1.Machine, error) {
+	allMachines, err := getMachines(ctx, crClient, clusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	var ownedMachines []*clusterv1.Machine
+	for i := range allMachines.Items {
+		m := allMachines.Items[i]
+		controllerRef := metav1.GetControllerOf(&m)
+		if controllerRef != nil && controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == kcp.Name {
+			ownedMachines = append(ownedMachines, &m)
+		}
+	}
+
+	return ownedMachines, nil
+}
+
+func getMachineNode(ctx context.Context, crClient client.Client, machine *clusterv1.Machine) (*corev1.Node, error) {
+	nodeRef := machine.Status.NodeRef
+	if nodeRef == nil {
+		return nil, nil
+	}
+
+	node := &corev1.Node{}
+	err := crClient.Get(
+		ctx,
+		types.NamespacedName{Name: nodeRef.Name},
+		node,
+	)
+	if err != nil {
+		if apierrors.IsNotFound(errors.Cause(err)) {
+			return nil, nil
```

Comment:
If the machine has a node ref, it means that we previously were able to match a node to the machine by provider ID. If the node is now not found, isn't that problematic?

Reply:
Possibly, I was mainly trying to safeguard against future use when a Machine might be in the process of Deleting and the Node is being deleted, or the Machine is in some type of a failed state and the cloud provider deleted the Node.

Since this is only used within the context of updating status, I wouldn't want to fail updating the status just because the backing Node that was referenced by a NodeRef is no longer present.

Reply:
SGTM